Mathlib/RingTheory/PowerSeries/Comp.lean

```diff
@@ -0,0 +1,274 @@
+/-
+Copyright (c) 2023 Richard M. Hill. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Richard M. Hill
+-/
+import Mathlib.RingTheory.PowerSeries.Basic
+/-!
+# Definitions
+
+Let `R` be a commutative semiring. Give two formal power series `f(X)` and `g(X)` with coefficients
+in `R`, their formal composition, when it exists, is the power series
+
+  `f ( g ( X ))= ∑ₙ fₙ * g^n`
+
+I.e the `d`-th coefficient of the composition is the sum
+
+  `∑ₙ fₙ * coeff R d (g ^ n)`.
+
+The formal composition exists when all of these sums have finite support. This happens for example
+when `f` is a polynomial, or when the constant term of `g` is nilpotent. There are also other cases
+where the composition is defined, although these cases are less easy to classify.
+
+In this file we define
+
+  `PowerSeries.hasComp` : a relation on `R⟦X⟧`, where `f.hasComp g` means that the formal
+                          composition of `f` and `g` exists.
+
+  `PowerSeries.comp`    : a binary operation on `R⟦X⟧`, where `f.comp g` is the formal composition
+                          in the case `f.hasComp g`, or zero otherwise.
+
+## Notation
+
+The operation `f.comp g` can also be written `f ∘ᶠ g`.
+
+
+## Main results
+
+  `add_hasComp` if `f.hasComp h` and `g.hasComp h` then `(f+g).hasComp h`.
+
+  `coe_hasComp` if `f` is a polynomial then `(f : R⟦X⟧).hasComp h`.
+
+  `hasComp_of_isNilpotent_constantCoeff` if the constant term of `g`
+                is nilpotent then `f.hasComp g`.
+
+  `add_comp` if `f.hasComp h` and `g.hasComp h` then `(f + g) ∘ᶠ h = f ∘ᶠ h + g ∘ᶠ h`.
+
+  `coe_comp_eq_aeval` if `f` is a polynomial then `f ∘ᶠ g = aeval g f`.
+-/
+
+open Nat Finset BigOperators Polynomial
+open scoped Classical
+
+namespace PowerSeries
+
+section CommutativeSemiring
+variable {R : Type*} [CommSemiring R]
+
+/--`f.hasComp g` states that the power series `g` may be substituted into
+the power series `f = ∑ₙ fₙ * Xⁿ` to give a new power series, whose `d`-coefficient is
+
+  `∑ₙ fₙ * coeff R d (g ^ n)`
+
+For the formal composition to make sense, we require that each of these sums
+has finite support. There are two common situations when `f.hasComp g`:
+either `f` could be a polynomial or the constant term of `g` could be zero.
+However, there are other intermediate cases if `R` is not an integral domain.
+-/
+def hasComp (f g : R⟦X⟧) : Prop := ∀ d, ∃ N, ∀ n, N ≤ n → (coeff R n f) * coeff R d (g^n) = 0
```

Comment:
The naming guide says that propositions should be `CamelCas`, so
```suggestion
def HasComp (f g : R⟦X⟧) : Prop := ∀ d, ∃ N, ∀ n, N ≤ n → coeff R n f * coeff R d (g^n) = 0
```

Reply:
thank you.